### ISSUE TITLE
Turborepo does not correctly invalidate cache when dependencies' source code changes.

### ISSUE THREAD

**Issue by VanTanev** [vercel/turborepo#1274]
### What version of Turborepo are you using?

1.2.9

### What package manager are you using / does the bug impact?

pnpm

### What operating system are you using?

Linux

### Describe the Bug

In a monorepo, turbo's cache does not invalidate when a dependency's source code changes and the task has the following definition:

```json
{
  "pipeline": {
    "cypress:run": {
      "outputs": []
    }
  }
}
```

However, it does correctly invalidate if we have this definition:
```json
{
  "pipeline": {
    "cypress:run": {
      "dependsOn": ["^cypress:run"],
      "outputs": []
    }
  }
}
```

Even if the dependency doesn't have the `cypress:run` script.

In fact, it also invalidates correctly if we put in any other valid pipeline, eg, `^build:docs`, even if the dependencies didn't have a `build:docs` command.

### Expected Behavior

When we have a pipeline that doesn't depend on a specific task, but it does depend on its dependencies invalidating, maybe we could denote it like this:
```json
{
  "pipeline": {
    "cypress:run": {
      "dependsOn": ["^"],
      "outputs": []
    }
  }
}
```

### To Reproduce
Hattip to @mattpocock: 
https://www.loom.com/share/08311d9b8d26486a8134d6e22dc24fa7

**Comment by mattpocock** [comment]
@VanTanev This is a regression, working as expected in `1.2.0`

**Comment by gsoltis** [comment]
I'm not sure that this is in fact the wrong behavior. If you've stated that there are no dependencies for this task (`dependsOn` is empty), then `turbo` should not invalidate when a package-level dependency changes. Consider, for example, a `lint` task. Typically to lint a package, you don't care about that package's dependencies. 

In your case, it sounds like you *do* care about package-level dependencies, much like a typical `build` would depend on `^build`. This means that your `dependsOn` cannot be empty. 

The reason it works with any `^` task, including non-existent ones, is that once you tell `turbo` that your task requires anything from your dependencies, `turbo` considers dependency changes to invalidate your task. 

That brings us to the question of "what should be in `dependsOn`"? This is going to depend on your setup. What is actually required for this task? What is it about a dependency change that affects how this task runs? Should a change to a dependency `README.md` trigger invalidation? What about source code? Test code? Once we can answer what you are using from your dependencies, we can move on to how to reference them in your task's dependencies.

**Comment by gsoltis** [comment]
I appreciate that it may be frustrating, but it appears that the behavior you were relying on was a bug rather than the [documented behavior](https://turborepo.org/docs/reference/configuration#dependson). As such, I don't consider it a regression. If we were to revert whichever PR fixed it, I would consider that a regression as it would no longer be possible to specify tasks which did _not_ depend on anything else (e.g. the `lint` example above).

I can, however, offer you a path forward, with a small caveat. I would define a task in `turbo.json` called something like `collect-source`, and have your `cypress:run` task depend on `^collect-source`. I would further add an [`inputs`](https://turborepo.org/docs/reference/configuration#inputs) block to your `collect-source` task and scope it to just the source files, or whatever files you want to invalidate your `cypress:run` task. 

The small caveat is a note about specifying `inputs`, please see #1241 to avoid being surprised by a future change in this area. If you'd rather wait on that one, leaving `inputs` off of the `collect-source` task will cause an invalidation on any change within the package.

**Comment by sppatel** [comment]
This is essentially what I discovered as a viable path forward - https://github.com/vercel/turborepo/issues/1084#issuecomment-1129157459.

**Comment by kgregory** [comment]
> I'm not sure that this is in fact the wrong behavior. If you've stated that there are no dependencies for this task (`dependsOn` is empty), then `turbo` should not invalidate when a package-level dependency changes. Consider, for example, a `lint` task. Typically to lint a package, you don't care about that package's dependencies.

Ran into an issue relevant to this comment and I thought it was worth sharing.  There's a [very common scenario](https://github.com/vercel/turborepo/tree/main/examples/basic) that can cause issues for linting if you don't care about the package's dependencies.

Scenario:

- Have a workspace that defines a common eslint config
- Have other workspaces that depend on the eslint config workspace
- Have a lint pipeline without a topological `dependsOn` and make sure you've run it successfully (all good, cached!)
- Modify the eslint config in a way that would result in lint errors for its consumers (add a new rule)
- Run the lint pipeline again and observe the cache hit for your consumers (...all good?  Sadly, no.)

The change to the common eslint config has caused lint errors, but our lint pipeline succeeds with a cache hit because it isn't considering the topological dependencies when creating hashes.

The same is true for a typecheck pipeline if you happen to use a common tsconfig package (or the types in an internal package change).

**Comment by nathanhammond** [comment]
@kgregory would you mind posting this as a new issue?

It's a good example of something we should address—with either documentation or code.

**Comment by nathanhammond** [comment]
I'm going to lock this thread, the canonical discussion on a path forward is over at #2855. If you have additional comments we can continue the conversation there!

In short, this behavior is intentional. We're exploring alternative ways to make it easier to be successful with it, but it's not a bug. Over the last year we have _significantly_ improved our documentation story around this as well: https://turbo.build/repo/docs/core-concepts/monorepos/task-dependencies#dependencies-outside-of-a-task
